nexus/src/db/datastore.rs

```diff
@@ -1626,94 +1551,35 @@ impl DataStore {
             .await
             .map_err(|e| {
                 public_error_from_diesel_pool(
                     e,
-                    ErrorHandler::NotFoundByLookup(
-                        ResourceType::Instance,
-                        LookupType::ById(*instance_id),
-                    ),
+                    ErrorHandler::NotFoundByResource(authz_instance),
                 )
             })?;
         Ok(())
     }
 
-    /// Fetches a `NetworkInterface` from the database and returns both the
-    /// database row and an [`authz::NetworkInterface`] for doing authz checks.
-    ///
-    /// See [`DataStore::organization_lookup_noauthz()`] for intended use cases
-    /// and caveats.
-    // TODO-security See the note on organization_lookup_noauthz().
-    async fn network_interface_lookup_noauthz(
-        &self,
-        authz_instance: &authz::Instance,
-        interface_name: &Name,
-    ) -> LookupResult<(authz::NetworkInterface, NetworkInterface)> {
-        use db::schema::network_interface::dsl;
-        dsl::network_interface
-            .filter(dsl::time_deleted.is_null())
-            .filter(dsl::instance_id.eq(authz_instance.id()))
-            .filter(dsl::name.eq(interface_name.clone()))
-            .select(NetworkInterface::as_select())
-            .first_async(self.pool())
-            .await
-            .map_err(|e| {
-                public_error_from_diesel_pool(
-                    e,
-                    ErrorHandler::NotFoundByLookup(
-                        ResourceType::NetworkInterface,
-                        LookupType::ByName(interface_name.as_str().to_owned()),
-                    ),
-                )
-            })
-            .map(|d| {
-                (
-                    authz_instance.child_generic(
-                        ResourceType::NetworkInterface,
-                        d.id(),
-                        LookupType::from(&interface_name.0),
-                    ),
-                    d,
-                )
-            })
-    }
-
-    /// Lookup a `NetworkInterface` by name and return the full database record,
-    /// along with an [`authz::NetworkInterface`] for subsequent authorization
-    /// checks.
-    pub async fn network_interface_fetch(
-        &self,
-        opctx: &OpContext,
-        authz_instance: &authz::Instance,
-        name: &Name,
-    ) -> LookupResult<(authz::NetworkInterface, NetworkInterface)> {
-        let (authz_interface, db_interface) =
-            self.network_interface_lookup_noauthz(authz_instance, name).await?;
-        opctx.authorize(authz::Action::Read, &authz_interface).await?;
-        Ok((authz_interface, db_interface))
-    }
-
     /// Delete a `NetworkInterface` attached to a provided instance.
     pub async fn instance_delete_network_interface(
         &self,
         opctx: &OpContext,
-        authz_instance: &authz::Instance,
-        interface_name: &Name,
+        authz_interface: &authz::NetworkInterface,
```

Comment:
This change looks great to me overall, but I have one question. We need the `authz::NetworkInterface` to delete it, which seems OK, but in the delete-all function we need the `authz::Instance` with modify permissions. It seems like they should both have the same permissions check. Is this the desired behavior? I'm actually not sure with this new lookup API if one can express permissions "along the path", or only at the leaf. Just a clarification, which should not block merging.

Reply:
Yeah, good catch.  First, this change doesn't alter the permissions required for the `instance_delete_network_interface()` path.  But it does alter the permissions required for the `instance_delete_all_network_interfaces` path.  Previously, this function did no authz check at all.  Now, it checks whether you're allowed to modify the Instance.  It would be more consistent with `instance_delete_network_interface()` if we checked whether you could delete each of the interfaces.

I see a few options:

1. Restore the previous behavior.  Pros: none.  Cons: these functions are still inconsistent in the required permissions, _and_ one of them does no access check.
1. Declare that this is okay as-is.  Pros: no work. Cons: the behavior is a little confusing. Note that this would not be observable today because both the "modify" and "delete" actions both map to the "modify" permission, and anybody who has modify on either of these necessarily has modify on the other.  This is only because we don't support roles more fine-grained than a Project.  In order for this confusing behavior to be observable, I think we'd need to support (a) either a role that distinguished between "modify" and "delete" or (b) roles that can be assigned as fine-grained as the _network interface_.  Even more, this function is currently only used by a saga during unwind.  So the exposure to the weird behavior seems pretty small.
1. Give up on network interfaces as discrete things with their own permissions and have all of their operations check permissions on the Instance.  Pros: consistent behavior from the user's perspective.  Cons: actually, the implementation of network interfaces would look different than all other resources.
1. Modify `instance_delete_all_network_interfaces()` to enumerate all the interfaces and authorize/delete them individually.  Pros: consistent behavior.  Cons: more work, plus this operation would become a lot more expensive.

Both (2) and (4) seem okay to me.  I'm inclined towards (2), especially given that we can also change this without impacting anything for quite some time if we decide this is an important case to clean up.  But I could also implement (4) if that seems a lot better!

Reply:
Thanks Dave, that's helpful. To be clear, I was definitely not advocating that we _do_ anything, so (2) sounds perfect!

I think I'm looking for clarity on when we pick one set of permissions over the other. Here it seems that it was mostly because there's no customer-visible distinction between Modify and Delete for these two resources (a customer has one iff they have the other), and the performance impact of (4) is pretty clear. I also realize that there are no roles that are more fine-grained than a Project, in which case the question might be a bit academic. But were there other considerations you had when making this change? And "no" is a fine answer too :)

Feel free to land this whenever, I'm just curious and not interested in gating anything.

Reply:
Here's how I think of it: if possible: each resource should have a corresponding `authz` type and we should use `opctx.authorized` with the action we're doing on that exact resource.  We should avoid doing permissions checks on a resource that's different from the one we're actually taking action on -- that sort of logic belongs in the policy file instead.  This case feels like a rationalized exception to me.  I'm not sure how generalizable it is.  As a thought experiment: suppose we _did_ care about precision here, and we also didn't want the expense of separately loading every child object and its roles.  Say we wanted to provide an operation like "wipe all the Projects in this Organization but keep the Organization", and we didn't want to have to load and authorize everything under the Organization.  We _could_ define a new type of action -- "wipe"? -- that does this, and that only requires permissions on the top-level resource.  That's not the cleanest thing because it means there are multiple permissions that allow you to do the same thing.  But it might be a reasonable tradeoff for the performance benefit.  We should generally avoid this if we can, I think.

I'm not sure how to generalize from any of this.

Reply:
Also, I'm interpreting your comments as a "+1"/approval of the change :)